fl4health/clients/apfl_client.py

```diff
@@ -1,255 +1,101 @@
+        self.local_optimizer: Optimizer
+
+    def is_start_of_local_training(self, step: int) -> bool:
+        return step == 0
+
+    def update_after_step(self, step: int) -> None:
+        if self.is_start_of_local_training(step) and self.model.adaptive_alpha:
+            self.model.update_alpha()
+
+    def split_optimizer(self, global_optimizer: Optimizer) -> Tuple[Optimizer, Optimizer]:
+        """
+        The optimizer from get_optimizer is for the entire APFLModule. We need one optimizer
+        for the local model and one optimizer for the global model.
+        """
+        global_optimizer.param_groups.clear()
+        global_optimizer.state.clear()
+        local_optimizer = copy.deepcopy(global_optimizer)
+
+        global_optimizer.add_param_group({"params": [p for p in self.model.global_model.parameters()]})
+        local_optimizer.add_param_group({"params": [p for p in self.model.local_model.parameters()]})
+        return global_optimizer, local_optimizer
+
+    def setup_client(self, config: Config) -> None:
+        """
+        Set dataloaders, optimizers, parameter exchangers and other attributes derived from these.
+        """
+        super().setup_client(config)
+
+        # Split optimizer from get_optimizer into two distinct optimizers
+        # One for local model and one for global model
+        global_optimizer, local_optimizer = self.split_optimizer(self.optimizer)
+        self.optimizer = global_optimizer
+        self.local_optimizer = local_optimizer
+
+    def train_step(self, input: torch.Tensor, target: torch.Tensor) -> Tuple[Losses, torch.Tensor]:
         # Mechanics of training loop follow from original implementation
         # https://github.com/MLOPTPSU/FedTorch/blob/main/fedtorch/comms/trainings/federated/apfl.py
 
         # Forward pass on global model and update global parameters
-        self.global_optimizer.zero_grad()
-        global_pred = self.model(input, personal=False)["global"]
+        self.optimizer.zero_grad()
+        global_pred = self.model.global_forward(input)
         global_loss = self.criterion(global_pred, target)
         global_loss.backward()
-        self.global_optimizer.step()
+        self.optimizer.step()
 
         # Make sure gradients are zero prior to foward passes of global and local model
         # to generate personalized predictions
         # NOTE: We zero the global optimizer grads because they are used (after the backward calculation below)
         # to update the scalar alpha (see update_alpha() where .grad is called.)
-        self.global_optimizer.zero_grad()
+        self.optimizer.zero_grad()
         self.local_optimizer.zero_grad()
 
         # Personal predictions are generated as a convex combination of the output
         # of local and global models
-        pred_dict = self.model(input, personal=True)
-        personal_pred, local_pred = pred_dict["personal"], pred_dict["local"]
+        personal_pred = self.predict(input)
```

Comment:
I think this might be another situation where that might be a little confusing to a user who doesn't know the internals of our APFL implementation. They might be asked to define a predict but know that there are a few "predict" modes for APFL. One thing we could do, is have predict take a config dictionary that defaults to `{}`. Then we could implement it here where it is something like
```
def predict(input: torch.Tensor, auxiliary_args: Dict[str, Scalar]) -> torch.Tensor:
    if auxiliary_args['global_forward']:
        return self.model.global_forward(input)
    else:
        return self.model(input)
```
Just a thought.

Reply:
Updated predict to return either torch tensor or dict of torch tensor if there is multiple prediction types. This ended up working nicely in having changing minimal code. Let me know if this is okay f462d69](https://github.com/VectorInstitute/FL4Health/pull/62/commits/f462d69c915f7f19a2d97d808e685d1d0ad39dd6)